astrokat/observe_main.py

```diff
@@ -537,10 +561,34 @@ def run_observation(opts, kat):
                         )
                     )
                     continue
+
             # TODO: setup of noise diode pattern should be moved to sessions
             #  so it happens in the line above
-            if "noise_diode" in obs_plan_params.keys():
-                noisediode.pattern(kat.array, session, obs_plan_params["noise_diode"])
+            if "noise_diode" in obs_plan_params:
+                nd_setup = obs_plan_params["noise_diode"]
+                nd_lead = None
+                if "lead_time" in nd_setup:
+                    nd_lead = nd_setup['lead_time']
+
+                # Set noise diode period to multiple of correlator integration time.
+                if not kat.array.dry_run:
+                    cbf_corr = session.cbf.correlator
+                    dump_period = cbf_corr.sensor.int_time.get_value()
+                else:
+                    dump_period = 0.5  # sec
+                user_logger.debug('DEBUG: Correlator integration time {} [sec]'
+                                  .format(dump_period))
+
+                if "cycle_len" in nd_setup:
+                    cycle_len_frac = -(nd_setup['cycle_len'] // -dump_period)
+                    nd_setup['cycle_len'] = cycle_len_frac * dump_period
```

Comment:
This part looks interesting.  It is a little different to the old method.  Seems like the new approach tends to result in cycles being one dump period longer.  What led to this change?

```python
In [1]: def old(cycle_len, dump_period): 
   ...:     return int(cycle_len / dump_period) * dump_period 
   ...:                                                                                                                                                                    

In [2]: def new(cycle_len, dump_period): 
   ...:     cycle_len_frac = -(cycle_len // -dump_period) 
   ...:     return cycle_len_frac * dump_period 
   ...:                                                                                                                                                                    

In [3]: for c, d in zip([0.9, 1.0, 1.1, 0.9, 1.0, 1.1], [0.5, 0.5, 0.5, 0.499975, 0.499975, 0.499975]): 
   ...:     print('cycle len {}, dump period {} => old: {}, new: {}'.format(c, d, old(c, d), new(c, d))) 
   ...:                                                                                                                                                                    
cycle len 0.9, dump period 0.5 => old: 0.5, new: 1.0
cycle len 1.0, dump period 0.5 => old: 1.0, new: 1.0
cycle len 1.1, dump period 0.5 => old: 1.0, new: 1.5
cycle len 0.9, dump period 0.499975 => old: 0.499975, new: 0.99995
cycle len 1.0, dump period 0.499975 => old: 0.99995, new: 1.499925
cycle len 1.1, dump period 0.499975 => old: 0.99995, new: 1.499925
```

Reply:
I am a little confused at my choice of `cycle_len_frac = -(cycle_len // -dump_period)`
The issue here is that I am calculating a `ceil` with this calculation, instead of the `floor` `cycle_len_frac = cycle_len // dump_period` that I want >:(

I updated the code to calculate the floor value

```
def old(cycle_len, dump_period):
    return int(cycle_len / dump_period) *dump_period

def new(cycle_len, dump_period):
    cycle_len_frac = cycle_len // dump_period
    return cycle_len_frac * dump_period

for c, d in zip([0.9, 1.0, 1.1, 0.9, 1.0, 1.1], [0.5, 0.5, 0.5, 0.499975, 0.499975, 0.499975]):
    print('cycle len {}, dump period {} => old: {}, new: {}'.format(c, d, old(c, d), new(c, d)))

cycle len 0.9, dump period 0.5 => old: 0.5, new: 0.5
cycle len 1.0, dump period 0.5 => old: 1.0, new: 1.0
cycle len 1.1, dump period 0.5 => old: 1.0, new: 1.0
cycle len 0.9, dump period 0.499975 => old: 0.499975, new: 0.499975
cycle len 1.0, dump period 0.499975 => old: 0.99995, new: 0.99995
cycle len 1.1, dump period 0.499975 => old: 0.99995, new: 0.99995
```

Reply:
@ajoubertza , @mserylak 
On the following comment, I do not like the fact that the "new" accidental method would have silently hid the fact that the cycle length was shorter than the correlator dump period.
I am not sure how the ND Tsys calibration works, but I would assume some consistency in output power, which will then depend on where you are in the correlator dump, or is it so fast that it is in every correlator dump? But then, how much in every correlator dump, given we do not know the actual correlator dump period until we run the observation and read the sensor?
This just feels unpredictable....

I would rather explicitly issue a warning or and error if this happen

Else, we should remove the `floor` calculation entirely -- since aligning with correlator dumps are then not required?

Reply:
Sorry, I know nothing about ND TSys.

If it could be of scientific value to run an observation with cycle time < dump period (so < 0.5 seconds, typically), then skipping the flooring in that case sounds good.  If it isn't scientifically useful, then raise an exception. 

Yes, the simulation and dry-runs might differ slightly from the real run, but it is an edge case.

Reply:
Mentally I evaluated all the options that could arise:
- Using correlator or beamformer data with `ND cycle len < dump rate`
There may be a possibility in engineering and integration tests where you may want to look at the fraction of power measured per correlator dump, compared to some threshold value or distribution calculated from the expected ND power contribution over the dump period.
But this seems like a lot of work for a test with little probability to give useful results.....

- BLUSE and F-engine full Nyquist data capture (nano-second dump rates)
Here the correlator dump rate does not matter, since you read the data directly out of the F-engines (time domain transient search)
And `ND cycle len < dump rate` is valid
But here you will want to keep the value provided by the user since now you do not know what the readout rate is -- it may be full Nyquist, of user defined averaging < correlator dump rate.

And herein lies the issue... If you use correlator/beamformer output, then you like to alignment to the dump rate, since you can predict which dumps will be affected
So if `ND cycle len >= dump rate` it makes sense to floor
But `ND cycle len >= dump rate` is not a determining condition for the use of correlator/beamformer data, it is merely an evaluation
The way to distinguish between the need to align with a correlator dump boundary vs "I do not care" is thus very ill defined. And will only be defined with more observational use.

So I am going to propose the following compromise:
IF `ND cycle len >= dump rate` apply the correlator boundary alignment (since this is known to be desired for correlator output)
ELSE apply user specified value
```
if (nd_setup['cycle_len'] >= dump_period):
    cycle_len_frac = nd_setup['cycle_len'] // dump_period
    nd_setup['cycle_len'] = cycle_len_frac * dump_period
```

What I am hoping this will achieve is all the basic functionality, which means that scientists can use it and give feedback on the refinement of needs.
Also, from what I know about the BL guys, is that they are very good at aligning data themselves and missing a dump here or there because of contamination does not seem to be to big an issue.

So I am making the least amount of assumptions, by only using the integer calculation in the case where I know it will still be useful and give a valid result -- thus there will not be any unexpected behaviour such as no noise diode signal.

Reply:
OK. If I want ND to flash at 0.089 s, the system should allow me to do it. Maybe issue a warning in the script log about non-standard ND operation requested?

Reply:
@mserylak yes, theoretically, you will now be allowed to set very short ND flash periods.
But it cannot be guaranteed to be against a sample edge though

Is it really a non-standard ND operation? It seems more a non-standard correlator implementation.....

Can we split the difference?
Since the observation script already issues a warning to tell the user that it updated the cycle length to the correlator dump period.
```
Set noise diode period to multiple of correlator dump period: cycle length = 16.0 [sec]
```
I added an `else` option that will update the warning for the `< dump_period` case to inform the user that the cycle length is less than the correlator dump period
```
Requested cycle length 0.1s < correlator dump period 0.5s, ND not aligned with dump edge
```

This way the user can ignore the warning if it does not matter, or
complain if the functionality is not as desired, and we have a new test case to update the development

Reply:
Just a reminder that the dump edges and the noise diode switching edges are not necessarily _aligned_.  This approach above for the `nd_setup['cycle_len'] >= dump_period` case will _synchronise_ the edges, but doesn't align them.  Alignment is a long outstanding issue (I was reminded of this old issue:  [SR-1637](https://skaafrica.atlassian.net/browse/SR-1637), see also its related issue) - it can be done but we need an exact timestamp for a dump edge in the current subarray build.

Attempt to explain via an example:
```
cycle length = 4 dump periods
period (duty cycle) = 75%

ASCII example:
dump period = 4 characters
cycle length = 4 * 4 = 16 chars
period on = 16 * .75 = 12 chars 
period off = 16 * .25 = 4 chars 
diode on:  x
diode off: _

correlator dump:     [ 0][ 1][ 2][ 3][ 4][ 5][ 6][ 7][ 8]
noise diode 1:      _xxxxxxxxxxxx____xxxxxxxxxxxx____xxxxxxxxxxxx____
noise diode 2:      __xxxxxxxxxxxx____xxxxxxxxxxxx____xxxxxxxxxxxx___
noise diode 3:      ___xxxxxxxxxxxx____xxxxxxxxxxxx____xxxxxxxxxxxx__
noise diode 4:      ____xxxxxxxxxxxx____xxxxxxxxxxxx____xxxxxxxxxxxx_
noise diode 5:      _____xxxxxxxxxxxx____xxxxxxxxxxxx____xxxxxxxxxxxx_
time:             |-------------------------------------------------------> 
```
Note:  the 5 noise diode rows are not different diodes in the same subarray, but
rather examples of 5 different time alignments that could occur across the
subarray for different noise diode pattern requests.  It depends what
actual time was used for turning on the noise source.

Reply:
You are correct @ajoubertza, and thank you very much for the detail explanation.
I was overstating the functionality.
I corrected the wording as you suggested below.

I assume more work will follow when we eventually get to the mythical ECP64

Reply:
@ajoubertza alignment will be working if you specify the same startup time on noise diodes in the array - that is under condition that it is at least 3+ seconds of lead time. I performed an extensive work to test that and proved that alignment is now functioning reliably on both L- and UHF-band receivers. Please see here: https://skaafrica.atlassian.net/browse/COMM-347.
@rubyvanrooyen I understand. In case of firing the noise diode at periods much shorter than correlator dump is non-standard for imaging but totally normal for beamformer. So, the solution you implemented (earning etc.) is sufficient for beamformer user needs.

Reply:
@mserylak Wow, that was very detailed analysis, and bug finding!  The analysis was based on F-engine data, right?  It shows that noise diodes are aligned with each other, but nothing about the alignment of the noise diode cycles with the correlator dump cycles.  My comment was about the alignment with correlator dump cycles.

Also, for that analysis the timestamps for turning on would also have been rounded up to the nearest integer (`np.ceil` in astrokat).  This PR removes that functionality.  Do you think the rounding up matters?

Reply:
@ajoubertza thanks. Yes, it was F-engine data and yes, at that time (and actually now as well) it did not (and still does not) matters for beamformer (BF) to have noise diodes aligned with correlator timestamps. When you use ND to calibrate pulsar observation you receive beamformer stream which is Nyquist sampled and then reduced (dedispersed, folded, integrated etc.). After it comes out sampled at ~us to ms timescales and you do not need long ND observation. If I was to wait for seconds just to pulse ND once that would be a waste of time in BF "world". For pulsar observation you need to flash ND for a minute or less at fast rate (e.g. Parkes uses ~11.123 Hz / ~0.89 ms period with 50% duty cycle) and integrate it to one profile. That is enough for pulsar calibration. Imaging requires longer because their sampling time is 1000s time longer (msec vs sec). They also want to keep the ND on as well.